jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/CrossOriginHandler.java

```diff
@@ -0,0 +1,423 @@
+                    LOG.debug("cross-origin request matches allowed origins: {} {}", request, getAllowedOriginPatterns());
+
+                if (preflight)
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("preflight cross-origin request {}", request);
+                    handlePreflightResponse(origins, response);
+                }
+                else
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("simple cross-origin request {}", request);
+                    handleSimpleResponse(origins, response);
+                }
+
+                if (timingOriginMatches(origins))
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("cross-origin request matches allowed timing origins: {} {}", request, getAllowedTimingOriginPatterns());
+                    response.getHeaders().put(HttpHeader.TIMING_ALLOW_ORIGIN, origins);
+                }
+            }
+
+            if (preflight && !isDeliverPreflightRequest())
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("preflight cross-origin request not delivered to child handler {}", request);
+                callback.succeeded();
+                return true;
+            }
+        }
+        return super.handle(request, response, callback);
+    }
+
+    private boolean originMatches(String origins)
+    {
+        if (anyOriginAllowed)
+            return true;
+        if (allowedOriginPatterns.isEmpty())
+            return false;
+        return originMatches(origins, allowedOriginPatterns);
+    }
+
+    private boolean timingOriginMatches(String origins)
+    {
+        if (anyTimingOriginAllowed)
+            return true;
+        if (allowedTimingOriginPatterns.isEmpty())
+            return false;
+        return originMatches(origins, allowedTimingOriginPatterns);
+    }
+
+    private boolean originMatches(String origins, Set<Pattern> allowedOriginPatterns)
+    {
+        for (String origin : origins.split(" "))
+        {
+            origin = origin.trim();
+            if (origin.isEmpty())
+                continue;
+            for (Pattern pattern : allowedOriginPatterns)
+            {
+                if (pattern.matcher(origin).matches())
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean isPreflight(Request request)
+    {
+        return HttpMethod.OPTIONS.is(request.getMethod()) && request.getHeaders().contains(HttpHeader.ACCESS_CONTROL_REQUEST_METHOD);
+    }
+
+    private void handlePreflightResponse(String origin, Response response)
+    {
+        HttpFields.Mutable headers = response.getHeaders();
+        headers.put(HttpHeader.ACCESS_CONTROL_ALLOW_ORIGIN, origin);
+        if (isAllowCredentials())
+            headers.put(HttpHeader.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
```

Comment:
This should be a shared `PreEncodedHttpField` rather than constructed every response.